gammapy/visualization/datasets.py

```diff
@@ -0,0 +1,177 @@
+        plot_kwargs = kwargs.copy()
+        plot_kwargs["facecolor"] = "None"
+        plot_kwargs.setdefault("edgecolor")
+        plot_kwargs.update(props)
+
+        dataset.counts_off.plot_region(ax, **plot_kwargs)
+
+        # create proxy artist for the custom legend
+        if legend:
+            handle = Patch(**plot_kwargs)
+            handles.append(handle)
+            labels.append(dataset.name)
+
+    if legend:
+        legend = ax.get_legend()
+        if legend:
+            handles = legend.legendHandles + handles
+            labels = [text.get_text() for text in legend.texts] + labels
+
+        handles = [(handle, handle) for handle in handles]
+        tuple_handler = HandlerTuple(ndivide=None, pad=0)
+
+        def patch_func(
+            legend, orig_handle, xdescent, ydescent, width, height, fontsize
+        ):
+            radius = width / 2
+            return CirclePolygon((radius - xdescent, height / 2 - ydescent), radius)
+
+        patch_handler = HandlerPatch(patch_func)
+
+        legend_kwargs.setdefault("handletextpad", 0.5)
+        legend_kwargs["handler_map"] = {Patch: patch_handler, tuple: tuple_handler}
+        ax.legend(handles, labels, **legend_kwargs)
+
+    return ax
+
+
+def plot_npred_signal(
+    dataset,
+    ax=None,
+    model_names=None,
+    region=None,
+    plot_background=True,
+    stack=False,
+    **kwargs,
+):
+    """
+    Plot the energy distribution of predicted counts of a selection of models assigned to a dataset.
+
+    Parameters
+    ----------
+    dataset : an instance of `~gammapy.datasets.dataset`
+        The dataset from which to plot the npred_signal.
+    ax : `~matplotlib.axes.Axes`
+        Axis object to plot on.
+    model_names : list of str
+        The list of models for which the npred_signal is plotted. Default is None.
+        If None, all models are considered.
+    region: `~regions.Region` or `~astropy.coordinates.SkyCoord`
+        Region used to reproject predicted counts. Default is None.
+        If None, use the full dataset geometry.
+    plot_background : bool
+        Whether to plot the background along with the other models.
+    stack : bool
+        Whether to sum the npred_signal of all the model.
+    **kwargs : dict
+        Keyword arguments to pass to `~gammapy.maps.RegionNDMap.plot`.
+
+
+    Returns
+    -------
+
+    axes : `~matplotlib.axes.Axes`
+        Axis object
+    """
+
+    npred_region = dataset.npred_signal(
+        model_names=model_names, stack=stack
+    ).to_region_nd_map(region)
+
+    if not stack:
+        npred_iter = npred_region.iter_by_axis("models")
+        names = npred_region.geom.axes["models"].center
+    else:
+        npred_iter = [npred_region]
+        names = ["stacked models"]
+
+    if ax is None:
+        fig = plt.figure()
+        axes = fig.add_subplot(111)
+    else:
+        axes = ax
+
+    for (npred, name) in zip(npred_iter, names):
```

Comment:
`RegionNDMap.plot()` already features the loop over the secondary axis. This is triggered by specifying the x-axis, like `RegionNDMap.plot(axis_name="energy")`. Please try it put and simplify the code here even further...